core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/StopStartUpscaleGetRecoveryCandidatesHandler.java

```diff
@@ -0,0 +1,187 @@
+
+            if (clusterHealthService.isClusterManagerRunning()) {
+                instanceMetadataWithServicesNotRunning = collectRecoveryCandidates(startedInstances, detailedHostStatuses,
+                        runtimeVersionService.getRuntimeVersion(stack.getClusterId()));
+            }
+
+            List<CloudInstance> instancesWithServicesNotRunning;
+
+            if (instanceMetadataWithServicesNotRunning.isEmpty()) {
+                instancesWithServicesNotRunning = Collections.emptyList();
+            } else {
+                instancesWithServicesNotRunning = instanceMetaDataToCloudInstanceConverter.convert(instanceMetadataWithServicesNotRunning, stack.getStack());
+                LOGGER.info("Collected {} upscale recovery candidates for hostGroup {}", instancesWithServicesNotRunning.size(),
+                        request.getHostGroupName());
+            }
+
+            StopStartUpscaleGetRecoveryCandidatesResult result =
+                    new StopStartUpscaleGetRecoveryCandidatesResult(request.getResourceId(), request, instancesWithServicesNotRunning,
+                            request.getAdjustment(), request.getHostGroupName());
+            notify(result, event);
+        } catch (Exception e) {
+            String errorMessage = String.format("Failed to collect upscale recovery candidates for stack: %s, proceeding with subsequent flow.",
+                    request.getResourceId());
+            LOGGER.warn(errorMessage, e);
+            StopStartUpscaleGetRecoveryCandidatesResult result = new StopStartUpscaleGetRecoveryCandidatesResult(request.getResourceId(), request,
+                    Collections.emptyList(), request.getAdjustment(), request.getHostGroupName());
+            notify(result, event);
+        }
+    }
+
+    private List<CloudInstance> collectStartedInstancesFromCloudProvider(CloudConnector connector, AuthenticatedContext ac,
+            List<CloudInstance> cloudInstances) {
+        List<CloudVmInstanceStatus> vmInstanceStatuses = connector.instances().checkWithoutRetry(ac, cloudInstances);
+        return vmInstanceStatuses.stream()
+                .filter(vm -> InstanceStatus.STARTED.equals(vm.getStatus()))
+                .map(CloudVmInstanceStatus::getCloudInstance)
+                .collect(Collectors.toList());
+    }
+
+    private AuthenticatedContext getAuthenticatedContext(StopStartUpscaleGetRecoveryCandidatesRequest request, CloudContext cloudContext,
+            CloudConnector connector) {
+        return connector.authentication().authenticate(cloudContext, request.getCloudCredential());
+    }
+
+    //CHECKSTYLE:OFF
+    private List<InstanceMetadataView> collectRecoveryCandidates(List<InstanceMetadataView> startedInstances, DetailedHostStatuses detailedHostStatuses,
+            Optional<String> runtimeVersion) {
+        boolean servicesHealthCheckAllowed = CMRepositoryVersionUtil.isCmServicesHealthCheckAllowed(runtimeVersion);
```

Comment:
This check needs to be in DetailedHost* system itself. Otherwise, very use to think that service health info is available even for the older versions, where it is not.